halo2/src/recursion/endoscale.rs

```diff
@@ -0,0 +1,67 @@
+//! Gadget for endoscaling.
+
+use ff::PrimeFieldBits;
+use halo2_gadgets::utilities::decompose_running_sum::be;
+use halo2_proofs::{
+    circuit::{AssignedCell, Layouter},
+    plonk::Error,
+};
+use pasta_curves::arithmetic::CurveAffine;
+
+mod chip;
+pub mod primitive;
+
+/// Instructions to map bitstrings to and from endoscalars.
+pub trait EndoscaleInstructions<C: CurveAffine>
+where
+    C::Base: PrimeFieldBits,
+{
+    /// Computes commitment (Alg 1) to a variable-length bitstring using the endoscaling
+    /// algorithm.
+    ///
+    /// The bitstring is decomposed into pairs of bits using a running sum outside of
+    /// this gadget.
+    #[allow(clippy::type_complexity)]
+    fn endoscale_base<L: Layouter<C::Base>, const NUM_BITS: usize, const NUM_WINDOWS: usize>(
```

Comment:
We possibly want `endoscale_fixed_base` (for public inputs) and `endoscale_var_base` (when multiplying by challenges).